File: app/src/main/java/com/roy/ui/FApps.java
package com.roy.ui;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.roy.R;
import com.roy.a.AppAdapter;
import com.roy.app.AppsSingleton;
import com.roy.itf.AppsInterface;
import com.roy.model.App;
import com.roy.sv.BroadcastReceivers;
import com.roy.util.UtilAppSorter;
import com.roy.util.UtilSettings;

import java.util.ArrayList;
import java.util.Objects;

import me.zhanghai.android.materialprogressbar.MaterialProgressBar;

public class FApps extends Fragment implements AppsInterface {
    RecyclerView rvApps;
    MaterialProgressBar progressBarApps;

    private UtilSettings utilSettings;
    private int indexScrolledItem;

    public FApps() {
    }

    public static FApps newInstance() {
        return new FApps();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.f_apps, container, false);
        utilSettings = new UtilSettings(getActivity());
        setupRecycler(Objects.requireNonNull(AppsSingleton.getInstance()).getApps());
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        setupViews(view);
    }

    private void setupViews(View view) {
        rvApps = view.findViewById(R.id.rvApps);
        progressBarApps = view.findViewById(R.id.progressBarApps);
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        if (getActivity() != null && getActivity() instanceof ASettings) {
            ((ASettings) getActivity()).setAppsInterface(this);
        }
    }

    private void sendEditAppsBroadcast() {
        if (getActivity() == null) {
            return;
        }
        Intent editAppsIntent = new Intent(getActivity(), BroadcastReceivers.AppsEditedReceiver.class);
        getActivity().sendBroadcast(editAppsIntent);
    }

    private void setupRecycler(ArrayList<App> apps) {
        if (getActivity() == null || apps.size() == 0) {
            return;
        }
        if (rvApps.getLayoutManager() != null) {
            indexScrolledItem = ((LinearLayoutManager) rvApps.getLayoutManager()).findFirstCompletelyVisibleItemPosition();
        }
        progressBarApps.setVisibility(View.INVISIBLE);
        rvApps.setVisibility(View.VISIBLE);
        AppAdapter mAppAdapter = new AppAdapter(getActivity(), apps);
        rvApps.setAdapter(mAppAdapter);
        rvApps.setLayoutManager(new GridLayoutManager(getActivity(), getResources().getInteger(R.integer.columns_apps)));
        rvApps.setItemAnimator(new DefaultItemAnimator());
        rvApps.scrollToPosition(indexScrolledItem);
        indexScrolledItem = 0;
    }

    @Override
    public void onDefaultsReset() {
        if (utilSettings.getSortType() != UtilAppSorter.SortType.values()[UtilSettings.DEFAULT_SORT_TYPE]) {
            utilSettings.save(UtilSettings.KEY_SORT_TYPE, UtilSettings.DEFAULT_SORT_TYPE);
            sendEditAppsBroadcast();
        }
    }

    @Override
    public void onAppsUpdated(ArrayList<App> apps) {
        setupRecycler(apps);
    }
}

File: app/src/main/java/com/mckimquyen/ui/FApps.kt
package com.mckimquyen.ui

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.view.isVisible
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.mckimquyen.R
import com.mckimquyen.a.AppAdapter
import com.mckimquyen.app.AppsSingleton.Companion.instance
import com.mckimquyen.enums.SortType
import com.mckimquyen.itf.AppsInterface
import com.mckimquyen.model.App
import com.mckimquyen.sv.BroadcastReceivers.AppsEditedReceiver
import com.mckimquyen.util.UtilSettings
import me.zhanghai.android.materialprogressbar.MaterialProgressBar

class FApps : Fragment(), AppsInterface {

    companion object {
        fun newInstance(): FApps {
            return FApps()
        }
    }

    private var rvApps: RecyclerView? = null
    private var progressBarApps: MaterialProgressBar? = null
    private var utilSettings: UtilSettings? = null
    private var indexScrolledItem = 0

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.f_apps, container, false)
        utilSettings = UtilSettings(requireContext())
        instance?.apps?.let {
            setupRecycler(it)
        }
        return view
    }

    override fun onViewCreated(
        view: View, savedInstanceState: Bundle?
    ) {
        super.onViewCreated(view, savedInstanceState)
        setupViews(view)
    }

    private fun setupViews(view: View) {
        rvApps = view.findViewById(R.id.rvApps)
        progressBarApps = view.findViewById(R.id.progressBarApps)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (activity != null && activity is ASettings) {
            (activity as ASettings?)?.setAppsInterface(this)
        }
    }

    private fun sendEditAppsBroadcast() {
//        if (activity == null) {
//            return
//        }
        val editAppsIntent = Intent(activity, AppsEditedReceiver::class.java)
        activity?.sendBroadcast(editAppsIntent)
    }

    private fun setupRecycler(apps: ArrayList<App>?) {
        if (activity == null || apps?.size == 0) {
            return
        }

        rvApps?.layoutManager?.let { lm ->
            indexScrolledItem = (lm as LinearLayoutManager).findFirstCompletelyVisibleItemPosition()
        }

        progressBarApps?.isVisible = false
        rvApps?.isVisible = true

        val appAdapter = AppAdapter(activity, apps)
        rvApps?.apply {
            adapter = appAdapter
            layoutManager =
                GridLayoutManager(activity, resources.getInteger(R.integer.columns_apps))
            itemAnimator = DefaultItemAnimator()
            scrollToPosition(indexScrolledItem)
        }
        indexScrolledItem = 0
    }

    override fun onDefaultsReset() {
        utilSettings?.let { us ->
            if (us.sortType != SortType.values()[UtilSettings.DEFAULT_SORT_TYPE]) {
                us.save(UtilSettings.KEY_SORT_TYPE, UtilSettings.DEFAULT_SORT_TYPE)
                sendEditAppsBroadcast()
            }
        }
    }

    override fun onAppsUpdated(apps: ArrayList<App>?) {
        setupRecycler(apps)
    }
}
